repository/remote_http_test.go

```diff
@@ -0,0 +1,160 @@
+	"net/http"
+	"strings"
+	"testing"
+)
+
+func Test_newHTTPRemote(t *testing.T) {
+	t.Run("Should return new httpRemote", func(t *testing.T) {
+		config := &core.Remote{}
+		creds := &core.Credentials{}
+
+		_, err := newHTTPRemote(config, creds)
+
+		require.NoError(t, err)
+	})
+}
+
+func Test_checkStatusCode(t *testing.T) {
+	t.Run("should return nil for HTTP 200", func(t *testing.T) {
+		mockResp := &http.Response{}
+		mockResp.Status = "200 OK"
+		mockResp.StatusCode = http.StatusOK
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(`{"status": "is well"}`))
+
+		// when
+		err := checkStatusCode(mockResp)
+
+		// then
+		require.NoError(t, err)
+	})
+
+	t.Run("should return error for HTTP statuses >= 300", func(t *testing.T) {
+		mockResp := &http.Response{}
+		mockResp.Status = "300 Whoopsie!"
+		mockResp.StatusCode = 300
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(`{"status": "I, uh, well... phew!"}`))
+
+		// when
+		err := checkStatusCode(mockResp)
+
+		// then
+		require.Error(t, err)
+		assert.Equal(t, err.Error(), "remote registry returns invalid status: 300 Whoopsie!: I, uh, well... phew!: (no error)")
+	})
+
+	t.Run("should return error for HTTP 400", func(t *testing.T) {
+		const errorBody = "Do not use v1 endpoint for v2 dogu creation. Use v2 endpoint instead."
+
+		mockResp := &http.Response{}
+		mockResp.Status = http.StatusText(http.StatusBadRequest)
+		mockResp.StatusCode = http.StatusBadRequest
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(fmt.Sprintf(`{"error": "%s"}`, errorBody)))
+
+		// when
+		err := checkStatusCode(mockResp)
+
+		// then
+		require.Error(t, err)
+		assert.Equal(t, err.Error(), "remote registry returns invalid status: Bad Request: 400: Do not use v1 endpoint for v2 dogu creation. Use v2 endpoint instead.")
+	})
+
+	t.Run("should return custom error for HTTP 401", func(t *testing.T) {
+		mockResp := &http.Response{}
+		mockResp.Status = http.StatusText(http.StatusUnauthorized)
+		mockResp.StatusCode = http.StatusUnauthorized
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(`{"status": "unauthorized"}`))
+
+		// when
+		err := checkStatusCode(mockResp)
+
+		// then
+		require.Error(t, err)
+		assert.True(t, commonerrors.IsUnauthorizedError(err))
+	})
+
+	t.Run("should return custom error for HTTP 403", func(t *testing.T) {
+		mockResp := &http.Response{}
+		mockResp.Status = http.StatusText(http.StatusForbidden)
+		mockResp.StatusCode = http.StatusForbidden
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(`{"status": "forbidden"}`))
+
+		// when
+		err := checkStatusCode(mockResp)
+
+		// then
+		require.Error(t, err)
+		assert.True(t, commonerrors.IsForbiddenError(err))
+	})
+
+	t.Run("should return custom error for HTTP 404", func(t *testing.T) {
+		mockResp := &http.Response{}
+		mockResp.Status = http.StatusText(http.StatusNotFound)
+		mockResp.StatusCode = http.StatusNotFound
+		mockResp.Body = ioutil.NopCloser(strings.NewReader(`{"status": "forbidden"}`))
```

Comment:
`{"status": "forbidden"}` copy paste? Shouldn't it be like "not found" or something?